### PR TITLE
fix(acvm): return false rather than panicking on invalid ECDSA signatures

### LINKED ISSUE #2755: ECDSA signature verification panics if provided invalid signature/public key
### Aim

I am trying to wrap up my [dark-safe circuit](https://github.com/colinnielsen/dark-safe/blob/main/circuits/src/main.nr#L26-L42):

I need a variable amount of items in an input array, so I'm using loop with some fixed `MAX_ITERATIONS`, then an `should_calculate` boolean in the input structs as a "no-op" flag.

There is an `if` block within the `for` loop, which looks for a `should_calculate` flag in the struct, and no-ops otherwise.

Screenshot 2023-09-19 at 7 30 11 PM

### Expected Behavior

The program should not panic

### Bug

Given a [Prover.toml](https://github.com/colinnielsen/dark-safe/blob/main/circuits/Prover.toml) with 4 valid inputs and with 4 "empty" structs,  running `nargo execute`, I see 4 `println` statements echo the array index, and "computing", then the program crashes at the blackbox-solver file.
Screenshot 2023-09-19 at 7 15 52 PM

Running the `nargo execute -p provers/Prover_full.toml` with a _full_ array [Prover_full.toml](https://github.com/colinnielsen/dark-safe/blob/main/circuits/Prover.toml) works as expected
Screenshot 2023-09-19 at 7 18 16 PM

### **NOTE:** 
- If I place the "empty" structs _first_ in the Prover.toml, it will crash on the first iteration in the no-op Screenshot 2023-09-19 at 7 21 16 PM
- I have also tried "flipping" the if statement, and putting the computation in the `else` branch. This makes no difference to either above case. The program crashes whenever it hits an empty struct in the loop.
- Interestingly enough, if I _only pass empty_ structs in via `nargo execute`, the program will fail on the first iteration

### To Reproduce

1. `git clone https://github.com/colinnielsen/dark-safe.git && cd dark-safe/circuits`
2. either `nargo execute`, `nargo execute -p provers/Prover_completely_empty.toml`, `nargo execute -p provers/Prover_full.toml`, or `nargo execute -p provers/Prover_empty_first.toml`

If you want to play with your own Prover.toml, (with a valid polynomial and array of signatures) run:
1. `yarn install`
2. `yarn build`
3. follow the cli prompt

### Installation Method

Binary

### Nargo Version

nargo 0.12.0 (git version hash: 13cc23bfe655eff52fa19c3f7e32f1c3cf09bd82, is dirty: false)

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves #2755
## Summary\*

Supercedes https://github.com/noir-lang/acvm/pull/557

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Review by kevaundray — APPROVED** [review]
LGTM - one nit, unwrap -> expect

**Comment by TomAFrench** [comment]
I think

```
 let pubkey = if pubkey.is_some().into() {
        pubkey.unwrap()
    } 
```

is pretty clear though.

File: acvm-repo/blackbox_solver/src/lib.rs [hunk 1]
```diff
@@ -117,15 +117,27 @@ fn verify_secp256k1_ecdsa_signature(
...
 
     let point = EncodedPoint::from_affine_coordinates(
         public_key_x_bytes.into(),
         public_key_y_bytes.into(),
         true,
     );
-    let pubkey = PublicKey::from_encoded_point(&point).unwrap();
 
+    let pubkey = PublicKey::from_encoded_point(&point);
+    let pubkey = if pubkey.is_some().into() {
```
Comment (jfecher) [inline acvm-repo/blackbox_solver/src/lib.rs:132]: It looks like `.into()` is converting a `bool` to a `bool` here. I'd prefer `if let` in general

Comment (TomAFrench) [reply acvm-repo/blackbox_solver/src/lib.rs:132]: We're not unwrapping an `Option` here but a [`CtOption`](https://docs.rs/subtle/latest/subtle/struct.CtOption.html#method.is_some) to `pubkey.is_some()` isn't a `bool` but `Choice`.

We can't do an `if let` as `CtOption` isn't an enum but a struct.

**Pull request merged**